### DISCUSSION TITLE
HOC(Higher-Order Component)란 무엇인가요?

### DISCUSSION THREAD

**Discussion by javascriptbiggosoo** [wanted-9th-3team/frontend-study#30] [프론트 기술 질문 공간]
사용 예시를 들어 설명해주세요.

**Comment by WooGie911** [comment]
### 고차 컴포넌트에 대한 경험이나 사전 지식이 없어 공식문서에서 정의만 가져왔습니다.

고차 컴포넌트(HOC, Higher Order Component)는 컴포넌트 로직을 재사용하기 위한 React의 고급 기술입니다. 
고차 컴포넌트(HOC)는 React API의 일부가 아니며, React의 구성적 특성에서 나오는 패턴입니다.

구체적으로, 고차 컴포넌트는 컴포넌트를 가져와 새 컴포넌트를 반환하는 함수입니다.

`const EnhancedComponent = higherOrderComponent(WrappedComponent)`

컴포넌트는 props를 UI로 변환하는 반면에, 고차 컴포넌트는 컴포넌트를 새로운 컴포넌트로 변환합니다.

고차 컴포넌트(HOC)는 Redux의 connect와 Relay의 createFragmentContainer와 같은 서드 파티 React 라이브러리에서 흔하게 볼 수 있습니다.

참고 : 리액트 공식문서

**Comment by michoball** [comment]
고차 컴포넌트는 반복되는 작업을 줄이고 컴포넌트의 재사용성을 높이기 위한 기술입니다. 
다른 컴포넌트를 인자로 받아서 새로운 컴포넌트를 리턴하는 구조로 되어있습니다. 

스타일링이나 로그인 인증, loader 표시 등을 할 때 등 쓰입니다. 

사용해본 적은 없지만, 로그인 인증으로 router를 차단할 때 사용하는 경우를 본 적이 있습니다.

**Comment by skyhanull** [comment]
고차 컴포넌트는 컴포넌트를 가져와 새 컴포넌트를 반환하는 함수입니다.

이부분은 잘 모르겠습니다 ㅠ

**Comment by BaekYuri** [comment]
고차 컴포넌트는 컴포넌트 로직을 재사용하기 위해 사용되며, 컴포넌트를 가져와 새로운 컴포넌트로 반환하는 함수입니다.
정적 메소드는 반드시 따로 복사해야 하고, render 메소드 안에서 고차 컴포넌트를 사용하면 안된다고 하는데 이 부분은 자세히는 모르겠습니다.
